### PR TITLE
chore(ids-apis): Use delegation types and providers to filter.

### PR DESCRIPTION
**Pull request by valurefugl**
https://www.notion.so/Use-the-new-delegation-type-to-authorize-instead-of-the-boolean-fields-178afd2080c74703b5306ce44df10900?pvs=4

## What

Use the new delegation types to authorize instead of the boolean fields.

## Why

So we can support more delegation types.

TODO: Also refactor the `delegation-resources.service.ts` and associated tests.

## Screenshots / Gifs

Attach Screenshots / Gifs to help reviewers understand the scope of the pull request

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Formatting passes locally with my changes
- [ ] I have rebased against main before asking for a review

## Summary by CodeRabbit

- **New Features**
  - Improved handling and support of different delegation types in the authentication services.
  - Added methods to dynamically retrieve provider information based on delegation types.

- **Bug Fixes**
  - Enhanced filtering logic for supported delegation types, ensuring accurate processing and scope retrieval.

- **Refactor**
  - Reordered and optimized import statements across various services.
  - Refactored methods to improve delegation handling and scope retrieval processes.

- **Chores**
  - Updated TypeScript types and added new DTOs for enhanced client and delegation management.

### CODE REVIEW

File: libs/auth-api-lib/src/lib/clients/clients.service.ts [hunk 1]
```diff
@@ -1,39 +1,41 @@
 /* eslint-disable  @typescript-eslint/no-explicit-any */
 import type { Logger } from '@island.is/logging'
-import { LOGGER_PROVIDER } from '@island.is/logging'
-import { NoContentException } from '@island.is/nest/problem'
 import { BadRequestException, Inject, Injectable } from '@nestjs/common'
 import { InjectModel } from '@nestjs/sequelize'
-import sha256 from 'crypto-js/sha256'
 import Base64 from 'crypto-js/enc-base64'
+import sha256 from 'crypto-js/sha256'
+import { AuthDelegationType } from 'delegation'
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply libs/auth-api-lib/src/lib/clients/clients.service.ts]: Fixed.

File: apps/services/auth/ids-api/src/app/delegations/delegations.controller.spec.ts [hunk 2]
```diff
@@ -1,15 +1,12 @@
 import { getModelToken } from '@nestjs/sequelize'
+import { AuthDelegationProvider, AuthDelegationType } from 'delegation'
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply apps/services/auth/ids-api/src/app/delegations/delegations.controller.spec.ts]: Fixed.

File: apps/services/auth/ids-api/src/app/delegations/delegations.controller.spec.ts [hunk 3]
```diff
@@ -27,44 +24,39 @@
...
-  createDelegation,
-  createDelegationModels,
   createDomain,
+  FixtureFactory,
 } from '@island.is/services/auth/testing'
+import {
+  AuthDelegationProvider,
+  AuthDelegationType,
+  getPersonalRepresentativeDelegationType,
+} from '@island.is/shared/types'
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply apps/services/auth/ids-api/src/app/delegations/delegations.controller.spec.ts:33]: Fixed.

File: apps/services/auth/ids-api/src/app/delegations/delegations.controller.ts [hunk 4]
```diff
@@ -101,48 +93,20 @@ export class DelegationsController {
...
-    let scopes = ([] as string[]).concat(...scopeSets)
-
     if (
-      scopes.length > 0 ||
-      delegationType.includes(AuthDelegationType.ProcurationHolder) ||
-      delegationType.includes(AuthDelegationType.LegalGuardian)
+      delegationType &&
+      delegationType.some(
+        (dt) => dt === AuthDelegationType.PersonalRepresentative,
+      )
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply apps/services/auth/ids-api/src/app/delegations/delegations.controller.ts]: Fixed.

File: apps/services/auth/ids-api/src/app/delegations/test/delegations-filters-types.ts [hunk 5]
```diff
@@ -20,7 +22,7 @@ export const user = createCurrentUser({
 
 export const legalGuardianScopes = ['lg1', 'lg2']
 export const procurationHolderScopes = ['ph1', 'ph2']
-export const customScopes = ['cu1', 'cu2']
+export const customScopes = ['cu1', 'cu2', 'cu-od1']
```
Comment (baering) [inline apps/services/auth/ids-api/src/app/delegations/test/delegations-filters-types.ts:25]: Should all these be hard coded?

File: apps/services/auth/ids-api/src/app/delegations/delegations.controller.spec.ts [hunk 6]
```diff
@@ -79,11 +70,28 @@ describe('DelegationsController', () => {
...
+      clientId: '@island.is/webapp',
+    })
+
+    client.allowedScopes = Object.values([
+      'scope/valid1',
+      'scope/valid2',
+      'scope/valid1and2',
+      'scope/unactiveType',
+      'scope/outdated',
+      'disabledScope',
```
Comment (baering) [inline apps/services/auth/ids-api/src/app/delegations/delegations.controller.spec.ts:83]: Should these not be in variables so they do not have to be referenced as strings in the tests?

File: apps/services/auth/ids-api/src/app/delegations/test/delegations-filters-types.ts [hunk 7]
```diff
@@ -62,23 +67,29 @@ export class TestCase {
...
+    return domains
   }
 
   get apiScopes(): CreateApiScope[] {
     return this.scopes.map((s) => ({
       name: s,
       description: s,
       displayName: s,
-      domainName: this.domainName,
+      domainName: s.includes('-od') ? this.otherDomainName : this.domainName,
```
Comment (baering) [inline apps/services/auth/ids-api/src/app/delegations/test/delegations-filters-types.ts:86]: Shouldn't `'-od'` be in a variable with a descriptive name?

File: apps/services/auth/ids-api/test/stubs/personalRepresentativeStubs.ts [hunk 8]
```diff
@@ -106,3 +109,10 @@ export default {
   getPRenabledApiScope,
   getScopePermission,
 }
+
+export const delegationTypes = [
+  AuthDelegationType.PersonalRepresentative + ':valid1',
+  AuthDelegationType.PersonalRepresentative + ':valid2',
+  AuthDelegationType.PersonalRepresentative + ':unactivated',
+  AuthDelegationType.PersonalRepresentative + ':outdated',
```
Comment (baering) [inline apps/services/auth/ids-api/test/stubs/personalRepresentativeStubs.ts:117]: More hard coded strings.

**Review by baering — APPROVED** [review]
LGTM except for some minor hard coded strings which I commented on.

File: apps/services/auth/ids-api/src/app/delegations/delegations.controller.spec.ts (hunk 6 above)
Comment (valurefugl) [reply apps/services/auth/ids-api/src/app/delegations/delegations.controller.spec.ts]: Fixed.

File: apps/services/auth/ids-api/src/app/delegations/test/delegations-filters-types.ts (hunk 5 above)
Comment (valurefugl) [reply apps/services/auth/ids-api/src/app/delegations/test/delegations-filters-types.ts]: Fixed.

File: apps/services/auth/ids-api/src/app/delegations/test/delegations-filters-types.ts (hunk 7 above)
Comment (valurefugl) [reply apps/services/auth/ids-api/src/app/delegations/test/delegations-filters-types.ts]: Fixed.

File: apps/services/auth/ids-api/test/stubs/personalRepresentativeStubs.ts (hunk 8 above)
Comment (valurefugl) [reply apps/services/auth/ids-api/test/stubs/personalRepresentativeStubs.ts:117]: Leaving as-is for now.

**Pull request merged**